exercises/practice/roman-numerals/.approaches/introduction.md

```diff
@@ -0,0 +1,204 @@
+    inverter = len(digits) - 1 
+
+    # translate decimal digits list to Roman numerals list
+    roman_digits = [table[inverter - i][d - 1] for (i, d) in enumerate(digits) if d != 0]
+
+    # convert the list of Roman numerals to a single string
+    return ''.join(roman_digits)
+```
+
+See [`table-lookup`][table-lookup] for details.
+
+
+## Loop over Romans approaches
+
+In this class of approaches we:
+- Create a mapping from Roman to Arabic numbers, in some suitable format.
+- Iterate nested loops, a `for` and a `while`, in either order.
+- At each step, append the largest possible Roman number to a list and substract the correspond value from the number being converted.
+- Join the list into a string and return it.
+Depending on the implementation, there may need to be a list-reverse step.
+
+This is one example using a dictionary:
+
+```python
+ROMAN = {1000: 'M', 900: 'CM', 500: 'D', 400: 'CD',
+         100: 'C', 90: 'XC', 50: 'L', 40: 'XL',
+         10: 'X', 9: 'IX', 5: 'V', 4: 'IV', 1: 'I'}
+
+def roman(number: int) -> str:
+    result = ''
+    while number:
+        for arabic in ROMAN.keys():
+            if number >= arabic: 
+                result += ROMAN[arabic]
+                number -= arabic
+                break
+    return result
+```
+
+There are a number of variants.
+See [`loop-over-romans`][loop-over-romans] for details.
+
+## Other approaches
+
+### Built-in methods
+
+Python has a package for pretty much everything, and Roman numerals are [no exception][roman-module].
+
+```python
+>>> import roman
+>>> roman.toRoman(23)
+'XXIII'
+>>> roman.fromRoman('MMDCCCLXXXVIII')
+2888
+```
+
+First it is necessary to install the package with `pip` or `conda`.
+Like most external packages, `roman` is not available in the Exercism test runner.
+
+This is the key part of the implementation on GitHub, which may look familiar:
+
+```python
+def toRoman(n):
+    result = ""
+    for numeral, integer in romanNumeralMap:
+        while n >= integer:
+            result += numeral
+            n -= integer
+    return result
+```
+
+The library function is a wrapper around a `loop-over-romans` approach.
+
+### Recursion
+
+This is a recursive version of the `loop-over-romans` approach, which only works in Python 3.10 and later:
+
+```python
+ARABIC_NUM = (1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1)
+ROMAN_NUM = ("M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I")
+
+def roman(number: int) -> str:
+    return roman_recur(number, 0, [])
+
+def roman_recur(num: int, idx: int, digits: list[str]):
+    match (num, idx, digits):
+        case [_, 13, digits]:
+            return ''.join(digits[::-1])
+        case [num, idx, digits] if num >= ARABIC_NUM[idx]:
+            return roman_recur(num - ARABIC_NUM[idx], idx, [ROMAN_NUM[idx],] + digits)
+        case [num, idx, digits]:
+            return roman_recur(num, idx + 1, digits)
```

Comment:
Oh!  LOVE THIS.  💖

Reply:
I wouldn't quite go that far, but I wanted to bring in pattern matching. Python as a Scala emulator...